### PR TITLE
Implement the remote data source request for topHeadlines

### LINKED ISSUE #12: Implement the remote data source
Implement the remote data source

The API we are going to use is this one.
https://newsapi.org/

For now we will focus on the endpoint TopHeadlines (/v2/top-headlines) and we will use the country parameter set to "us".
i.e.: https://newsapi.org/v2/top-headlines?country=us&apiKey=XYZ

More customisation will be added at a later date like specifying the **country**, **category** or **sources** 

More info can be found here: https://newsapi.org/docs/endpoints/top-headlines

### PR DESCRIPTION
**Pull request by Diogo00dev**
# Short Description
Resolves #12 
Implement the remote data source

# What changes in the code?
Added logic to actually make the network request to the news api to get the topHeadlines.
Test the request

# Did you test?

- [x] Yes
- [ ] No

### CODE REVIEW

File: flutter/news/lib/features/frontpage/data/datasource/articles_remote_data_source.dart [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  Future<Result<List<ArticleModel>>> topHeadLines() async {
+    var url = Uri.https(Constants.baseUrl, Constants.topHeadlinesEndpoint, {'country': 'us'});
+
+    var response = await client.get(url, headers: {'X-Api-Key': Constants.newsApiKey});
+    if (response.statusCode == 200) {
+      return Result.success(
+          BaseNewsResponseModel.fromJson(json.decode(response.body)).articles
+              as List<ArticleModel>);
+    } else {
+      throw ServerFailure("Server failure");
```
Comment (juankysoriano) [inline flutter/news/lib/features/frontpage/data/datasource/articles_remote_data_source.dart]: 🛑 

Here we should be returning a `Result.failure(ServerFailure)`

File: flutter/news/lib/features/frontpage/data/models/article_model.dart [hunk 2]
```diff
@@ -4,13 +4,13 @@ import 'package:news/features/frontpage/domain/entities/article.dart';
...
   ArticleModel(
       {required SourceModel source,
-      required String author,
-      required String title,
-      required String description,
-      required String url,
-      required String urlToImage,
-      required String publishedAt,
-      required String content})
+      required String? author,
```
Comment (juankysoriano) [inline flutter/news/lib/features/frontpage/data/models/article_model.dart:7]: are all this fields really optional?

File: flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      BaseNewsResponseModel.fromJson(json.decode(fixture("articles.json")))
+          .articles as List<ArticleModel>);
+
+  setUp(() {
+    mockHttpClient = MockClient();
+    remoteDataSource = ArticlesRemoteDataSource(client: mockHttpClient);
+  });
+
+  test(
+    'Given Get request is done When doing the request Then the endpoint '
```
Comment (juankysoriano) [inline flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart]: Having `GIVEN, WHEN, and THEN` in capital letters helps readability, I would go for consistency with other tests

File: flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    'being used should be /v2/top-headlines',
+    () async {
+      // arrange
```
Comment (juankysoriano) [inline flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart]: :) I am sure you wouldn't be adding this comments after reading `clean code`, 

There is a chapter about comments and Uncle Bob mentions there `Don't add obvious noise.`

**Comment by juankysoriano** [comment]
@Diogo00dev let's pair on this, there are a few improvements I want to explore

File: flutter/news/lib/features/frontpage/data/datasource/articles_remote_data_source.dart (hunk 1 above)
Comment (Diogo00dev) [reply flutter/news/lib/features/frontpage/data/datasource/articles_remote_data_source.dart]: updated on [34d2f99](https://github.com/novoda/multiplatform-playground/pull/15/commits/34d2f99e34fbfc03733e882fdcc758565e92553f)

File: flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart (hunk 3 above)
Comment (Diogo00dev) [reply flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart]: 100% agree

updated on [34d2f99](https://github.com/novoda/multiplatform-playground/pull/15/commits/34d2f99e34fbfc03733e882fdcc758565e92553f)

File: flutter/news/lib/features/frontpage/data/models/article_model.dart (hunk 2 above)
Comment (Diogo00dev) [reply flutter/news/lib/features/frontpage/data/models/article_model.dart:7]: did some more testing on the api and removed some unnecessary` ?`

updated on [34d2f99](https://github.com/novoda/multiplatform-playground/pull/15/commits/34d2f99e34fbfc03733e882fdcc758565e92553f)

File: flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart (hunk 4 above)
Comment (Diogo00dev) [reply flutter/news/test/features/frontpage/data/datasource/articles_remote_data_source_test.dart]: agree, have them on my custom template for tests and forgot to remove.

updated on [34d2f99](https://github.com/novoda/multiplatform-playground/pull/15/commits/34d2f99e34fbfc03733e882fdcc758565e92553f)

**Review by juankysoriano — APPROVED** [review]

**Pull request merged**